### ISSUE TITLE
approach to adding a11y to sun buttons

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#633]
I just completed conversion of sun buttons to ES6 class for phetsims/tasks#1044.  

One of the things that I noticed while doing this conversion is a lack of symmetry/coverage in how a11y features have been added to sun buttons.

Examples:

* RoundButtonView has `this.focusHighlight` and `options.focusHighlightDilation`.  `RectangularButtonView` has neither.

* RoundButtonView adds listener `updatePDOMEnabled` to `buttonModel.enabledProperty` to set the 'aria-disabled' attribute.  `RectangularButtonView` does nothing like that.

* RectangularButtonView has method `a11yClick`. RoundButtonView has no such method.

There are more examples, but I think that can get the discussion started. 

Some questions:

* Why is there a discrepancy in how a11y is implemented for round versus rectangle buttons? Is there something about their shape that results in different a11y needs?

* What has been the approach to adding a11y to sun buttons? Is there a holistic approach or plan? Or historically have specific buttons been made to work based on the needs of specific sims?

* What approach does PhET want to use going forward?

Labeling for discussion at developer meeting. @jessegreenberg @zepumph feel free to comment in advance.

**Comment by jessegreenberg** [comment]
> What has been the approach to adding a11y to sun buttons? Is there a holistic approach or plan?

Yes - and this is what we are striving for. The majority of a11y code for sun buttons is in ButtonModel and PressListener and any sun button using `ButtonModel.createListener` will be accessible. What is mentioned in this list are very minor differences in my opinion, but they are certainly things that could be cleaned up. To respond:

> RoundButtonView has this.focusHighlight and options.focusHighlightDilation

These are specific to round buttons. But I also don't see `focusHighlightDilation` used anywhere and I think it should be removed.

> RoundButtonView adds listener updatePDOMEnabled to buttonModel.enabledProperty to set the 'aria-disabled' attribute. RectangularButtonView does nothing like that.

Yes, this should be in RectangularButtonView as well. `aria-disabled` is also set in EnabledNode, it seems like RoundButtonView and RectangularButtonView should be using this instead. That is part of https://github.com/phetsims/sun/issues/585. Ill add a listener to RectangularButtonView until that is done.

> RectangularButtonView has method a11yClick. RoundButtonView has no such method.

Yes, adding this to RoundButtonView as well so the API matches makes sense to me.

**Comment by zepumph** [comment]
We discussed this briefly at developer meeting today. It seems like @jessegreenberg explained and cleaned things up well above. The discrepancies were largely in the view. We run a11y predominantly off of ButtonModel's created PressListeners. 

@ariel-phet also discussed the benefit of cleaning up a11y code that was likely added in more of a "prototype" phase of that project.

@pixelzoom please review, and bring back to dev meeting if you have further thoughts or questions.

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/sun/issues/639, it was determined that ButtonModel `a11yClickingProperty` is vestigial.   So while I appreciate that the specific problems that I mentioned above have been addressed, perhaps this issue is not ready to close yet.   Have sun buttons been given a thorough "going over", to verify that they are up-to-date with current a11y practices, and that a11y has been applied in a way that provides good symmetry/coverage ?

**Comment by pixelzoom** [comment]
On my very next task in sun buttons (reviewing https://github.com/phetsims/sun/issues/635), I ran into something in ButtonModel. In https://github.com/phetsims/sun/issues/639, it was determined that ButtonModel `a11yClickingProperty` is vestigial.   

So while I appreciate that the specific problems that I mentioned above have been addressed, perhaps this issue is not ready to close yet.   Have sun buttons been given a thorough "going over", to verify that they are up-to-date with current a11y practices, and that a11y has been applied in a way that provides good symmetry/coverage ?

**Comment by zepumph** [comment]
Over in https://github.com/phetsims/sun/issues/639#issuecomment-715611674 @jessegreenberg and I cleaned up some of the a11y clicking logic in sun buttons. Please review that, and feel free to comment on if you think that a11y could still be better in sun buttons.

**Comment by pixelzoom** [comment]
I don't really understand #639, but I'll take your word for it.   But that is still a specific issue, and I have yet to receive an answer to the general issue:

> Have sun buttons been given a thorough "going over", to verify that they are up-to-date with current a11y practices, and that a11y has been applied in a way that provides good symmetry/coverage ?

**Comment by zepumph** [comment]
I'm going to summarize the current state of sun buttons as it pertains to accessibility.

>RoundButtonView has this.focusHighlight and options.focusHighlightDilation. RectangularButtonView has neither.

focusHighlightDilation is no longer here, and the round-specific focusHighlight is only in the RoundButton correctly. The default focusHighlight is a rectangle.

>RoundButtonView adds listener updatePDOMEnabled to buttonModel.enabledProperty to set the 'aria-disabled' attribute. RectangularButtonView does nothing like that.

This is now factored out to ParallelDOM generally.

>RectangularButtonView has method a11yClick. RoundButtonView has no such method.

This has been factored out to ButtonNode, and is used consistently in both visual button implementations.

>Why is there a discrepancy in how a11y is implemented for round versus rectangle buttons? Is there something about their shape that results in different a11y needs?

The only discrepancy now is the focusHighlight which mimics the visuals.


>What has been the approach to adding a11y to sun buttons? Is there a holistic approach or plan? Or historically have specific buttons been made to work based on the needs of specific sims?

We are constantly improving. In general buttons are pretty simple, and the needs are pretty well defined and managed with what is in master now. Thanks for issues like this to make things even more streamlined!


I'm ready to close this. @pixelzoom feel free to do so after review.

**Comment by pixelzoom** [comment]
👍🏻 Closing
